File: JavaScript/320-Generalized Abbreviation.js
/**
 * @param {string} word
 * @return {string[]}
 */
var generateAbbreviations = function(word) {
    let res = [];
    let dfs = (index, number, curr) =>{
        if (index === word.length){
            if (number > 0) curr += number.toString();
            res.push(curr);
            return 
        }
        dfs(index+1, number+1, curr)
        dfs(index+1, 0, curr + (number > 0 ? number.toString() : "") + word[index])
    }
    dfs(0, 0, "");
    return res
};

File: C++/320-Generalized Abbreviation.cpp
// O(N*2^N)
class Solution {
public:
    vector<string> res;
    int N;
    void dfs(int ind, string& word, string curr) {
        if (ind == N) {
            res.push_back(curr);
            return;
        }
        // starts with a char;
        dfs(ind+1, word, curr + word[ind]);

        // starts with a digit
        for (int j=ind; j<N; j++) {
            string num = to_string(j-ind+1);
            if (j+1 < N) {
                num += word[j+1];
                dfs(j+2, word, curr + num);
            } else {
                dfs(j+1, word, curr + num);
            }
        }
    }
    vector<string> generateAbbreviations(string word) {
        this->N = word.size();
        dfs(0, word, "");
        return res;
    }
};